packages/jest-cli/src/cli/runCLI.js

```diff
@@ -28,57 +29,64 @@ const watch = require('../watch');
 
 const VERSION = require('../../package.json').version;
 
-module.exports = (
+module.exports = async (
   argv: Object,
-  root: Path,
+  roots: Array<Path>,
   onComplete: (results: ?AggregatedResult) => void,
 ) => {
   const realFs = require('fs');
   const fs = require('graceful-fs');
   fs.gracefulify(realFs);
 
   const pipe = argv.json ? process.stderr : process.stdout;
-  argv = argv || {};
   if (argv.version) {
     pipe.write(`v${VERSION}\n`);
     onComplete && onComplete();
     return;
   }
 
-  const _run = async ({config, hasDeprecationWarnings}) => {
+  const _run = async (
+    configs: Array<{config: Config, hasDeprecationWarnings: boolean}>,
+  ) => {
     if (argv.debug) {
-      logDebugMessages(config, pipe);
+      // TODO: https://github.com/facebook/jest/issues/3288
+      logDebugMessages(configs[0].config, pipe);
     }
 
-    createDirectory(config.cacheDirectory);
-    const hasteMapInstance = Runtime.createHasteMap(config, {
-      console: new Console(pipe, pipe),
-      maxWorkers: getMaxWorkers(argv),
-      resetCache: !config.cache,
-      watch: config.watch,
-    });
+    const hasteMapInstances = Array(configs.length);
+    const contexts = await Promise.all(
```

Comment:
following our meeting discussion: we can put both `sharedConfig` and `projectConfig` in the context. i think